Adds a Rsync recipe, basis for performing repository-related tasks locally

johnny-bit opened the pull request:
# Rsync recipe

### Installing

```php
// deploy.php

require 'vendor/deployphp/recipes/recipes/rsync.php';
```

### Configuration options

- **rsync_excludes** *(optional)*: accepts a *array* with files/dirs to be excluded from sending to server
- **rsync_user** *(optional)*: accepts a *string* with username to server
- **rsync_local_release_dir** *(optional)*: accepts a *string* with dirname where temporary repository cloning should take place before being sent to server

```php
// deploy.php

set('rsync_excludes', [
    '.git',
    'deployer_release',
    'releases',
    'deploy.php',
]);

set('rsync_user', 'server_user');

set('rsync_local_release_dir', '/tmp');
```

### Tasks

- `deploy:local_release` Creates local release directory
- `deploy:update_code` overrides standard `deploy:update_code` to update code locally instead of remotelly
- `deploy:rsync` perorms rsync from local release to remote

### Suggested Usage

This recipe performs all repository-related tasks locally, so the best way to use it, would be to use this instead of `common.php` recipe.

For using composer or other tools, You'd need to override `deploy:vendors` task and plug it in `deploy` chain.

gordalina commented:
Hey, nice recipe!

I have only a couple: current convention for configuration variables is that a recipe uses only one keyword, which is the same name as the recipe itself. The other is that you should not provide a default user.

## Configuration (example)

```php
+set('rsync', [
    'excludes' => [
        '.git',
        'deployer_release',
        'releases',
        'deploy.php',
    ],
    'user' => 'server_user',
    'local_release_dir' => '/tmp'
]);
```

## User (example)

```php
if (!get('rsync.user')) {
    throw new \RuntimeException("You need to specify a rsync.user");
}
```

johnny-bit commented:
Thanks for suggestions! Changes are applied in new commit, however, now without specifying username instead of bailing out, we should try to connect without username, as sometimes this is acceptable (whether usernames are configured in ssh config or are same on destination host as on source).

elfet commented:
Great! Cool recipe! :beer:

elfet commented:
Does this recipe works then you trying to deploy multiple servers in parallel?

johnny-bit commented:
Wait a bit before merging that. I have some very nice additions piled up.

technically it SHOULD work... however we have process when we rsync our dev
code to per-dev pre-staging environment, then code goes to repo, then goes
to stating and then goes to one final production. I don't really know how
would this work in parallel, bacause I never had parallels deploy scenario.

2015-06-11 12:46 GMT+02:00 Anton Medvedev <notifications@github.com>:

> Does this recipe works then you trying to deploy multiple servers in
> parallel?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/deployphp/recipes/pull/10#issuecomment-111083051>.
>



-- 
Pozdrawiam,
Hubert Kowalski

elfet commented:
@johnny-bit try to deploy it on one server to different `deploy_path` and run with `dep deploy -p`.

johnny-bit commented:
It seems that it SHOULD work, but parallel deploy in my environment simply
hangs. I even provisioned different hosts to be able to use multiple hosts
and to no avail. It properly synces to multiple hosts in multiple
locations, but I can't get parallel deploy to work. It's either about my
fresh php5.6 or that I don't have suggested modules (i try to have on dev
host only that what's available on staging & production)

2015-06-11 13:53 GMT+02:00 Anton Medvedev <notifications@github.com>:

> @johnny-bit <https://github.com/johnny-bit> try to deploy it on one
> server to different deploy_path and run with dep deploy -p.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/deployphp/recipes/pull/10#issuecomment-111103620>.
>



-- 
Pozdrawiam,
Hubert Kowalski

johnny-bit commented:
Okayy - last commit is rather big, but allows great flexibility when it comes to rsync. Now deploys should be multiple-host-safe with pararell deployments being probably ok.

elfet commented:
Nice work! :) Awesome!

gordalina commented on docs/rsync.md:
````diff
@@ -0,0 +1,51 @@
+```php
+// deploy.php
+
+set('rsync',[
````
Please use the same block you have in `rsync.php` (the full default configuration)

gordalina commented on recipes/rsync.php:
```diff
@@ -0,0 +1,153 @@
+  
+})->desc('Rsync local->remote');
+
+task('deploy', [
```
Should a recipe override this? @elfet 
I understand that this way, its plug and play, but for custom deploy scripts, automatically overriding certain tasks (e.g.: deploy:update_code) might cause unwanted effects.

This is a major discussion as it's the first time its happening in recipes and will create a standard for future ones.

If no default tasks can be overwritten, the end-user must explicitely add rsync tasks e.g.:

```php
task('deploy', [
// ...
'deploy:rsync:local_release',
'deploy:rsync:update_code',
// ...
]);
```

elfet replied:
I think too, what 3rd-party recipes should has no effect on.
Imagine i has Symfony app based on `recipe/symfony.php`, but i want to use rsync. Now if i require this recipe - my deploy will be broken. Sad =(.

gordalina commented on recipes/rsync.php:
```diff
@@ -0,0 +1,153 @@
+    'cleanup',
+])->desc('Deploy your project');
+
+after('deploy', 'success');
```
Also, this should not exist.
Current recipes only expose tasks, but not define behaviour.

Lets talk about this

elfet replied:
I'm agree with @gordalina. Should only expose tasks.

johnny-bit commented:
So to sum up recent commit:

- Fixed documentation as @gordalina suggested
- recipe now only exposes one task: rsync

However in making this change I needed to make some decisions due to some time constraints today. Usually there are only one server and one source that would be constant... but what if for one server You'd depoly to /var/www and second to /var/www/vhosts/client_id/site_name/httpdocs/app? What if say with one command You'd want to deploy to one server from devel dir but to second server from locally built release? so I added 2 env vars: `rsync_src` and `rsync_dest` that can deal with it.

Now @elfet and @gordalina - how do You like it?

elfet commented:
> You'd depoly to /var/www and second to /var/www/vhosts/client_id/site_name/httpdocs/app? What if say with one command You'd want to deploy to one server from devel dir but to second server from locally built release? so I added 2 env vars: rsync_src and rsync_dest that can deal with it.

Common recipe based on env var `deploy_path`, and you should use it too.

johnny-bit commented:
Ummm... No.

`deploy_path` fails short. 

Reasons:

- `deploy_path` is for example /var/www/vhosts/site. However since we're doing releasey stuff, actuall content is in /var/www/vhosts/site/current. Well, that's a symlink, so it's actually in /var/www/vhosts/site/releases/[release_date]. If we were to rsync stuff using delete flag, then actual conent would go in /var/www/vhost/site but due to `delete` flag so popular with rsync, current and releases would vanish. So You'd need to use **`release_path`**
- We're doing al releasey stuff locally, for dev previews. But using convention we're specyfing `deploy_path` to be actual path to what we're syncing. Since there's no releasey stuff going on on remote, using previous `release_path` would yeld incorrect path for rsync. So... Need to use `deploy_path`.
- Say you'd need to rsync some misc files after deploy, but with nice rsync capabilities and you're DevOps spec. Files You're syncing are relative to `deploy_path` but are not kept in versioning system, maybe they're sensitive (happened couple times to me when deploying domain-registrar stuff). So neither `deploy_path` nor `release_path` are okay in this case. Actually maybe You could use `{{deploy_path}}/../../secure/devops/{{server.host}}/` and set it as `rsync_dest` and be done with it?

You see: I'm all for that super nice and super nifty flexibility. That way You can allow anybody using this recipe to have one command rsync tool that can do anything normal rsync can without having to specify advanced magic every time.

Look at it this way: Due to way of how I set it, it's most flexible rsync tool that would allow anybody to take advantage of rsync in any way they could come up with :smile_cat:

elfet commented:
Correct, what about using just `release_path`? :) Do not try to cover 100% usage on rsync =)
User can implement this it self. But having `deploy:update_code` which using rsync will be nice :+1: 
@gordalina ?

johnny-bit commented:
Well: my point is exactly not covering 100% possible usage, but instead
allow user to implement everything rather easily :) Using `release_path`
might be ok, but it would mess bot non-deployment workflow and
deployment-relative workflow :)

Trade-offs everywhere, but I personally really hate situation when I need
to create tremendous workaround just because creator hardcoded something
instead of allowing config.

Thoughts @gordalina ?

gordalina commented on docs/rsync.md:
```diff
@@ -0,0 +1,81 @@
+  'options' => ['delete'],
+]);
+
+env('rsync_src', __DIR__);
```
I would add here that these are the defaults and that it is possible to do more than one rsync deploy

gordalina commented:
Looks good to me! I kinda like that it is flexible enough to do other rsync stuff, not only the repository itself. 
@elfet do you have any other issues?

johnny-bit commented:
I've split config & environment variables. How does it look now?

gordalina commented:
@elfet do you have any concerns about this, or can it be merged?

elfet commented:
Everything look cool! Lets merge it! 
PS we a need some page on deployer.org with info about this recipes with docs.

gordalina commented:
@elfet that's on #8 but I haven't had the time for that, maybe @johnny-bit is interested in doing that?
